src/main/java/com/jfrog/ide/idea/scan/YarnScanner.java

```diff
@@ -63,5 +71,65 @@ protected AbstractInspection getInspectionTool() {
     protected PackageManagerType getPackageManagerType() {
         return PackageManagerType.YARN;
     }
+
+    /**
+     * Builds a map of package name to versions out of a set of <package-name>:<version> Strings.
+     *
+     * @param packages - A set of packages in the format of 'package-name:version'.
+     * @return - A map of package name to a set of versions.
+     */
+    Map<String, Set<String>> getPackageNameToVersionsMap(Set<String> packages) {
+        Map<String, Set<String>> packageNameToVersions = new HashMap<>();
+        for (String fullNamePackage : CollectionUtils.emptyIfNull(packages)) {
+            String[] packageSplit = StringUtils.split(fullNamePackage, ":");
+            String packageName = packageSplit[0];
+            String packageVersion = packageSplit[1];
+            packageNameToVersions.putIfAbsent(packageName, new HashSet<>());
+            packageNameToVersions.get(packageName).add(packageVersion);
+        }
+        return packageNameToVersions;
+    }
+
+    /**
+     * Builds the impact graph for each given vulnerable dependencies.
+     * The impact graph is built by running 'yarn why <package>' command, making it different from other package managers.
+     *
+     * @param vulnerableDependencies - The vulnerable dependencies to build the impact graph for.
+     *                               The key is the package name and version, and the value is the dependency node.
+     * @param depTree                - The whole dependency tree (not just vulnerable dependencies) that was generated earlier.
+     * @return - The impact graph attached to package.json DescriptorFileTreeNode
+     */
+    @Override
+    protected List<FileTreeNode> walkDepTree(Map<String, DependencyNode> vulnerableDependencies, DepTree depTree) throws IOException {
```

Comment:
* The method appears to be long and contains nested loops, making it complex. Consider breaking down the functionality into smaller, more focused methods. Each method should ideally perform a single logical operation.
* The method name `walkDepTree` does more than just walk, doesn't fully capture its purpose. Consider renaming it to reflect its functionality more explicitly.
* The method returns a List<FileTreeNode>, but internally, it seems to always return a list with a single element - the descriptorNode. It might be more intuitive to directly return the descriptorNode instead of encapsulating it within a list.